salt/pillar/__init__.py

```diff
@@ -273,83 +273,158 @@ def __init__(self, opts, grains, minion_id, saltenv, ext=None, functions=None,
+            self.ext_cache = salt.utils.cache.CacheFactory.factory(
+                    self.opts['ext_pillar_cache_backend'],
+                    self.opts['ext_pillar_cache_ttl'],
+                    minion_cache_path=self._minion_cache_path('ext_pillar_cache', minion_id))
+
+    def _minion_cache_path(self, cache_type, minion_id):
         '''
         Return the path to the cache file for the minion.
 
         Used only for disk-based backends
         '''
-        return os.path.join(self.opts['cachedir'], 'pillar_cache', minion_id)
+        return os.path.join(self.opts['cachedir'], cache_type, minion_id)
 
-    def fetch_pillar(self):
+    def fetch_pillar(self, fetch_pillar=True, fetch_ext_pillar=True):
         '''
         In the event of a cache miss, we need to incur the overhead of caching
         a new pillar.
         '''
-        log.debug('Pillar cache getting external pillar with ext: %s', self.ext)
+        ext = None
+        msg = 'Pillar cache getting external pillar with ext: {0}'.format(ext)
+        if fetch_ext_pillar is True:
+            ext = self.ext
+            msg = 'Ext_pillar cache getting external pillar with ext: {0}'.format(ext)
+        log.debug(msg)
         fresh_pillar = Pillar(self.opts,
                               self.grains,
                               self.minion_id,
                               self.saltenv,
-                              ext=self.ext,
+                              ext=ext,
                               functions=self.functions,
                               pillar_override=self.pillar_override,
-                              pillarenv=self.pillarenv)
+                              pillarenv=self.pillarenv,
+                              fetch_pillar=fetch_pillar,
+                              fetch_ext_pillar=fetch_ext_pillar)
         return fresh_pillar.compile_pillar()
 
     def compile_pillar(self, *args, **kwargs):  # Will likely just be pillar_dirs
-        log.debug('Scanning pillar cache for information about minion %s and pillarenv %s', self.minion_id, self.pillarenv)
-        log.debug('Scanning cache: %s', self.cache._dict)
-        # Check the cache!
-        if self.minion_id in self.cache:  # Keyed by minion_id
-            # TODO Compare grains, etc?
-            if self.pillarenv in self.cache[self.minion_id]:
-                # We have a cache hit! Send it back.
-                log.debug('Pillar cache hit for minion %s and pillarenv %s', self.minion_id, self.pillarenv)
-                return self.cache[self.minion_id][self.pillarenv]
+        pillar_ret = {}
+        ext_pillar_ret = {}
+        if self.cache:
+            log.debug('Scanning pillar cache for information about minion %s and pillarenv %s',
+                      self.minion_id, self.pillarenv)
+            log.debug('Scanning cache: %s', self.cache._dict)
+            # Check the cache!
+            if self.minion_id in self.cache:  # Keyed by minion_id
```

Comment:
This is getting to be a bit nested. Would it make sense at this point to break these out and refactor into new methods?